### PR TITLE
User/arudell/improve fault tolerance for requests

### PR DESCRIPTION
**Pull request by arudell**
# Description
Summary of changes:
- Fix issue with build validation function where it would fail build process if there was an empty file
- Added new functions
    - `Get-SdnRole`: used to determine the role of a given computer
    - `Get-UserInput`: used to capture input from operator
    - `Invoke-RestMethodWithRetry`: new function wrapper around `Invoke-RestMethod` that has a retry operation if `-Retry` is specified. Includes ability to handle `[System.Management.Automation.PSCredential]::Empty`
    - `Invoke-WebRequestWithRetry`: new function wrapper around `Invoke-WebRequest` that has a retry operation if `-Retry` is specified. Includes ability to handle `[System.Management.Automation.PSCredential]::Empty`
- Updated functions other functions to use the improved utility functions

# Change type
- [x] Bug fix (non-breaking change)
- [ ] Code style update (formatting, local variables)
- [x] New Feature (non-breaking change that adds new functionality without impacting existing)
- [ ] Breaking change (fix or feature that may cause functionality impact)
- [ ] Other

# Checklist:
- [x] My code follows the style and contribution guidelines of this project.
- [x] I have tested and validated my code changes.

### CODE REVIEW

File: src/modules/Common/private/Get-SdnRole.ps1 [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        [Parameter(Mandatory = $true)]
+        [System.String]$ComputerName
+    )
+
+    if ($null -eq $Global:SdnDiagnostics.EnvironmentInfo) {
+        throw New-Object System.NullReferenceException("Unable to enumerate data from EnvironmentInfo")
+    }
+
+    foreach ($role in ($Global:SdnDiagnostics.EnvironmentInfo.Keys | Where-Object {$_ -iin $Global:SdnDiagnostics.Config.Keys})) {
+        if ($ComputerName -iin $Global:SdnDiagnostics.EnvironmentInfo[$role]) {
```
Comment (luyaof) [inline src/modules/Common/private/Get-SdnRole.ps1:16]: Will this handle $ComputerName in both ShortName and FQDN format?

File: src/modules/Utilities/private/Invoke-WebRequestWithRetry.ps1 [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    if ($Credential -ne [System.Management.Automation.PSCredential]::Empty -and $null -ne $Credential) {
+        $params.Add('Credential', $Credential)
+    }
+    else {
+        $params.Add('UseDefaultCredentials', $true)
+    }
+
+    $retryIntervalInSeconds = 10
```
Comment (luyaof) [inline src/modules/Utilities/private/Invoke-WebRequestWithRetry.ps1:65]: Would it be useful if we make RetryInterval and MaxRetry as non-mandatory parameters with default value?

File: src/modules/Common/private/Get-SdnRole.ps1 (hunk 1 above)
Comment (arudell) [reply src/modules/Common/private/Get-SdnRole.ps1:16]: Good catch. I need to add this to the logic detection to handle FQDN vs just NetBIOS.

File: src/modules/Utilities/private/Invoke-WebRequestWithRetry.ps1 (hunk 2 above)
Comment (arudell) [reply src/modules/Utilities/private/Invoke-WebRequestWithRetry.ps1]: Added `-MaxRetry` and `-RetryIntervalInSeconds` as parameters for both Invoke methods.

File: src/modules/Common/private/Get-SdnRole.ps1 (hunk 1 above)
Comment (arudell) [reply src/modules/Common/private/Get-SdnRole.ps1]: Pushed a new code change to handle this.

**Review by luyaof — APPROVED** [review]

**Pull request merged**